metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala

```diff
@@ -121,6 +125,207 @@ final class BloopServers(
+      bloopGlobalJsonFilePath: AbsolutePath,
+      bloopCreatedByMetalsFilePath: AbsolutePath,
+      requestedBloopJvmProperties: List[String],
+      maybeJavaHome: Option[String],
+      reconnect: () => Future[BuildChange]
+  ): Future[Unit] = {
+    writeJVMPropertiesToBloopGlobalJsonFile(
+      bloopGlobalJsonFilePath,
+      bloopCreatedByMetalsFilePath,
+      requestedBloopJvmProperties,
+      maybeJavaHome
+    ) match {
+      case Failure(exception) => Future.failed(exception)
+      case Success(_) =>
+        languageClient
+          .showMessageRequest(
+            Messages.BloopJvmPropertiesChange.params()
+          )
+          .asScala
+          .flatMap {
+            case messageActionItem
+                if messageActionItem == Messages.BloopJvmPropertiesChange.reconnect =>
+              shutdownServer()
+              reconnect().ignoreValue
+            case _ =>
+              Future.successful(())
+          }
+    }
+
+  }
+
+  private def getBloopGlobalJsonFilePath(): Option[AbsolutePath] =
+    Try {
+      AbsolutePath(
+        Paths
+          .get(System.getProperty("user.home"))
+          .resolve(".bloop/bloop.json")
+      )
+    }.toOption
+
+  private def getBloopCreatedByMetalsFilePath(): Option[AbsolutePath] =
+    Try {
+      AbsolutePath(
+        Paths
+          .get(System.getProperty("user.home"))
+          .resolve(".bloop/created_by_metals.lock")
+      )
+    }.toOption
+
+  private def getBloopGlobalJsonLastModifiedTime(
+      bloopGlobalJsonFilePath: AbsolutePath
+  ): Long =
+    Try {
+      bloopGlobalJsonFilePath.toNIO.toFile.lastModified()
+    }.getOrElse(0)
+
+  /**
+   * First we check if the user requested to update the Bloop JVM properties through the extension.
+   * <p>If so, we also check if the Bloop's Global Json file exists and if it was pre-modified by the user.
+   * <p>Then, through consultation with the user through appropriate dialogues we decide if we should
+   * <ul>
+   * <li>overwrite the contents of the Bloop Global Json file with the requested JVM properties and Metal's JavaHome variables; and then restart the Bloop server</li>
+   * <li>or alternatively, leave things untouched</li>
+   * </ul>
+   *
+   * @param maybeRequestedBloopJvmProperties Bloop JVM Properties requested through the Metals Extension settings
+   * @param maybeRunningBloopJvmProperties
+   * @param maybeJavaHome                    Metals' `javaHome`, which Bloop should also preferebly use
+   * @param reconnect                        function to connect back to the build server.
+   * @return `Future.successful` if the purpose is achieved or `Future.failure` if a problem occured such as lacking enough permissions to open or write to files
+   */
+  def ensureDesiredJvmSettings(
+      maybeRequestedBloopJvmProperties: Option[List[String]],
+      maybeRunningBloopJvmProperties: Option[List[String]],
+      maybeJavaHome: Option[String],
```

Comment:
Wait a minute…
Why are there two `BloopJvmProperties` variables in here? I think I might have added it twice by mistake.
Also, frankly adding `maybe` really helps myself in writing code… So can it stay?

Reply:
> Also, frankly adding maybe really helps myself in writing code… So can it stay?

Of course :) it was just a suggestion. It comes down to personal preference, and I see in some other places we do the same, so no worries. Just one tip is to not rely too much on it, because sooner or later you'll hit on something called `maybe*` and it won't be an `Option`.

Reply:
Thanks for the tip!